### PR TITLE
provision.localhost: Add prepare functionality

### PR DESCRIPTION
**Pull request by thrix**
Adding 'ansible' and 'shell' steps to local provisioner.

Adding a few tests for the provisioner along the way :)

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

### CODE REVIEW

File: tmt/steps/provision/__init__.py [hunk 1]
```diff
@@ -8,36 +8,57 @@
...
 
 
 class Provision(tmt.steps.Step):
     """ Provision step """
-
-    # Default implementation for provision is a virtual machine
+    name = 'provision'
+
+    # supported provisioners are not loaded automatically, import them and map them in how_map
+    how_map = {
```
Comment (pvalena) [inline tmt/steps/provision/__init__.py]: How about a missing one from `how_map`? That should simply go to `ProvisionVagrant` as one can create their own `Providers` (vagrant calls them that) and they should simply run [without any modifications in ProvisionVagrant](https://github.com/psss/tmt/blob/master/tmt/steps/provision/vagrant.py#L275).

File: tmt/steps/provision/__init__.py [hunk 2]
```diff
@@ -8,36 +8,57 @@
...
+
+    def _check_data(self):
+        """ Validate input data """
+
+        # if not specified, use 'virtual' provisioner as default
+        for data in self.data:
+            how = data['how']
+            # is how supported?
+            if how not in self.how_map:
+                raise tmt.utils.SpecificationError("How '{}' in plan '{}' is not implemented".format(how, self.plan))
```
Comment (pvalena) [inline tmt/steps/provision/__init__.py]: I think there's no way to forward-check whether vagrant supports it, so we should simply input all the non-localhost `how`s to Vagrant. Please see https://github.com/psss/tmt/pull/43#issuecomment-553502265.

File: tmt/steps/provision/localhost.py [hunk 3]
```diff
@@ -1,7 +1,36 @@
...
+
+    def _prepare_ansible(self, what):
+        """ Run ansible on localhost """
+        # note: we expect playbooks are placed relatively to the current directory
+        self.run(f'ansible-playbook -c local -i localhost, {what}', cwd=os.getcwd())
+
+    def _prepare_shell(self, what):
+        """ Run ansible on localhost """
+        # note: we expect playbooks are placed relatively to the current directory
+        self.run(what, cwd=os.getcwd())
```
Comment (pvalena) [inline tmt/steps/provision/localhost.py]: @psss isn't it like the `tmt` looks for git root and operates on it? But maybe I've misunderstood. If not, it is implemented wrong for ProvisionVagrant then, as it relies on the playbook being in `discover`.

**Comment by pvalena** [comment]
Haven't checked the tests, but it looks good ! :))

I'll probably copy the tests and make them test ProvisionVagrant. ;)

File: tmt/steps/provision/__init__.py (hunk 1 above)
Comment (thrix) [reply tmt/steps/provision/__init__.py]: We can try it :)

File: tmt/steps/provision/__init__.py (hunk 2 above)
Comment (thrix) [reply tmt/steps/provision/__init__.py]: Agreed, ok. Will route everything other via vagrant.

**Comment by thrix** [comment]
> Haven't checked the tests, but it looks good ! :))
> 
> I'll probably copy the tests and make them test ProvisionVagrant. ;)

Yeah, we can just extend these tests wit Vagrant

File: tmt/steps/provision/localhost.py (hunk 3 above)
Comment (psss) [reply tmt/steps/provision/localhost.py]: Yes, all paths should be relative to the metadata tree root. Will fix this.

**Comment by psss** [comment]
Thanks for the patch. I've addressed comments from @pvalena and merged into master.

**Pull request merged**